Cleanup asserts & throws - replace with invariants - dir util/s* (part1)

sonodtt opened the pull request:
Cleanup of asserts - replaced with invariants.

Removed all unstructured `throw(...)` statements, replacing with `INVARIANT`, `PRECONDITION`, etc. where appropriate.

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -488,13 +488,13 @@ bool simplify_exprt::simplify_typecast(exprt &expr)
         {
-          expr=from_integer(0, expr_type);
-          assert(expr.is_not_nil());
+          expr = from_integer(0, expr_type);
```
Unrelated change

sonodtt replied:
removed.

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -1373,7 +1373,7 @@ bool simplify_exprt::simplify_update(exprt &expr)
         component_number(component_name);
 
-      assert(number<value_ptr->operands().size());
+      CHECK_RETURN(number < value_ptr->operands().size());
```
This code should be rewritten to use `get_component`.

sonodtt replied:
Refactored.

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -1409,7 +1409,9 @@ bool simplify_exprt::simplify_object(exprt &expr)
-    assert(expr.operands().size()==1);
+    DATA_INVARIANT(
+      expr.operands().size() == 1,
+      "typecasts must have exactly one argument");
```
Introduce `const typecast_exprt &tc = to_typecast_expr(expr);` above and change the code to use `tc` with its proper API. Let's not duplicate consistency checks on expressions.

sonodtt replied:
Done.

tautschnig commented on src/util/simplify_expr_boolean.cpp:
```diff
@@ -245,7 +245,9 @@ bool simplify_exprt::simplify_not(exprt &expr)
-    assert(op.operands().size()==2);
+    DATA_INVARIANT(
+      op.operands().size() == 2,
+      "exists expression has two parameters");
```
Please introduce `to_exists_expr` which should do this sort of check and use it here.

tautschnig commented on src/util/simplify_expr_floatbv.cpp:
```diff
@@ -142,7 +141,9 @@ bool simplify_exprt::simplify_floatbv_typecast(exprt &expr)
-  assert(expr.operands().size()==2);
+  PRECONDITION(expr.id() == ID_floatbv_typecast);
+  DATA_INVARIANT(
+    expr.operands().size() == 2, "float typecast has two operands");
```
This function should take a `floatbv_typecast_exprt` as a parameter and should not be doing those checks in here.

hannes-steffenhagen-diffblue replied:
I agree that'd be much cleaner, but note that currently all of the `simplify_XXX` work like this (this is also because many of the simplify functions can take many different types of expressions). We could change that but if we were to do that I'd change all of them at once and that's out of scope for this PR IMHO (if we just changed the one the code would be quite inconsistent)

tautschnig replied:
> [...] currently all of the [...]

Not quite true - there is sort of a gradual migration happening. Take a look at the declarations in `simplify_expr_class.h`.

hannes-steffenhagen-diffblue replied:
Hm... I don't really like that. It ends up violating structural invariants to still use the input expression as an out parameter, because most simplifications will "return" a different type than they were given... I find the idea of returning something that's _not_ a `plus_exprt` through a `plus_exprt` reference a bit disturbing.

tautschnig replied:
See https://github.com/diffblue/cbmc/pull/2910#discussion_r215971488.

hannes-steffenhagen-diffblue replied:
@tautschnig Can you take a look at https://github.com/diffblue/cbmc/pull/2939 and tell me if that's roughly in line with what you were thinking? Would you be OK with going ahead with this PR and doing further refactorings in one or more separate PRs?

sonodtt replied:
@tautschnig  I'm going to split this part out - and make a note for it to be considered in the context of #2939  and your remarks on https://github.com/diffblue/cbmc/pull/2910#discussion_r215971488

see 
https://github.com/diffblue/cbmc/pull/2939#issuecomment-424610272 that documents this

tautschnig replied:
Sounds good!

tautschnig reviewed (changes requested):
Let's try to do a bit more than the very mechanical cleanup.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: a2c876d16fb60b0652ab1c5f41d56998f3ebdbb5).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: a2c876d16fb60b0652ab1c5f41d56998f3ebdbb5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/83831905
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 9c79c23d13caf6732ae461fdb93065e10fbc95f5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84019963

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: b0b44fade96e6536db88ee41d7437087fb5b627d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85534606
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 42f660e92c2f59f17a8bb6a1ba484cc8b8eec7fe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85532856

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -1365,17 +1365,12 @@ bool simplify_exprt::simplify_update(exprt &expr)
-      value_ptr=&value_ptr->operands()[number];
+      auto &designator_as_struct_expr = to_struct_expr(*value_ptr);
+      value_ptr = &designator_as_struct_expr.component(component_name, ns);
+      CHECK_RETURN(value_ptr->id() != ID_nil);
```
Use `value_ptr->is_not_nil()`

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -1365,17 +1365,12 @@ bool simplify_exprt::simplify_update(exprt &expr)
-
-      if(!to_struct_type(value_ptr_type).
-         has_component(component_name))
+      const struct_typet &value_ptr_struct_type=to_struct_type(value_ptr_type);
```
Nit pick: whitespace around `=`

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -1407,18 +1402,13 @@ bool simplify_exprt::simplify_object(exprt &expr)
-      expr.operands().size() == 1,
-      "typecasts must have exactly one argument");
+    auto const& typecast_expr = to_typecast_expr(expr);
+    const typet &op_type=ns.follow(typecast_expr.op().type());
```
Space around `=`

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -1429,10 +1419,10 @@ bool simplify_exprt::simplify_object(exprt &expr)
 
-      exprt tmp=expr.op0();
-      if(tmp.id()==ID_plus && tmp.operands().size()==2)
+      exprt casted_expr=typecast_expr.op();
```
`const exprt &casted_expr = typecast_expr.op();`

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -1429,10 +1419,10 @@ bool simplify_exprt::simplify_object(exprt &expr)
-      exprt tmp=expr.op0();
-      if(tmp.id()==ID_plus && tmp.operands().size()==2)
+      exprt casted_expr=typecast_expr.op();
+      if(casted_expr.id()==ID_plus && casted_expr.operands().size()==2)
```
Add spacing around `==`

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -1429,10 +1419,10 @@ bool simplify_exprt::simplify_object(exprt &expr)
+      if(casted_expr.id()==ID_plus && casted_expr.operands().size()==2)
       {
-        exprt cand=tmp.op0().id()==ID_typecast?tmp.op0():tmp.op1();
+        exprt cand=casted_expr.op0().id()==ID_typecast?casted_expr.op0():casted_expr.op1();
```
Spacing

tautschnig commented on src/util/simplify_expr_boolean.cpp:
```diff
@@ -213,15 +213,12 @@ bool simplify_exprt::simplify_not(exprt &expr)
-    simplify_node(expr.op1());
+    auto const& op_as_exists = to_exists_expr(op);
+    forall_exprt rewritten_op(op_as_exists.symbol()
+      , op_as_exists.where());
```
```
forall_exprt rewritten_op(
  op_as_exists.symbol(), not_exprt(op_as_exists.where()));
```
and remove the line below

tautschnig commented on src/util/simplify_expr.cpp:
```diff
@@ -1365,7 +1365,8 @@ bool simplify_exprt::simplify_update(exprt &expr)
         e.get(ID_component_name);
-      const struct_typet &value_ptr_struct_type=to_struct_type(value_ptr_type);
+      const struct_typet &value_ptr_struct_type =
+        to_struct_type(value_ptr_type);
```
Please squash with the previous commit, the clang-format changes are restricted to code that you have introduced in the preceding commit.

Commit:
Cleanup asserts & throws - replace with invariants

Deprecated c-style asserts in favour of cbmc invariants.
Some minor refactoring that was relevant to the cleanup.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 8734b32c850d0269f023929cd691f1333d84ab0d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86134560
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: cba81d933c1a9afbb5c834b2e27e6f9f2718b1bc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86174685

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 6c0ab09e74e6b6b577b6b4252e1cd51a56b00dfb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86178555

hannes-steffenhagen-diffblue committed:
Refactor simplify_floatbv_typecast

* Use to_XXX_expr functions instead of asserts to make sure invariants are not
  violated
* Use named accessors instead of op0, op1 etc
* Rename variables (e.g. "casted_expr" instead of "op0") for readability

hannes-steffenhagen-diffblue commented:
@tautschnig I believe all of your comments (aside from changing the function signatures to be more strongly typed which will happen in another PR as previously discussed) have been addressed now.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: a168b24b8da1a95938a4df6fc65510347fc51e42).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86316058
